_posts/2019-12-20-testowanie-frontendu-asynchronicznego.md

```diff
@@ -0,0 +1,100 @@
+---
+layout:    post
+title:     Testowanie frontendu - Cz. 4: Testy jednostkowe kodu działającego asynchronicznie
+published: true
+date:      2019-12-20 08:00:00 +0100
+author:    amarszalek
+tags:
+    - Jasmine
+    - Angular
+    - async unit testing
+    - unit test
+    - komponent
+---
+## Wprowadzenie
+Kończąc serię dotyczącą testowania komponentów Angularowych przy pomocy Jasmine, chciałbym poruszyć temat testów kodu wykonywanego asynchronicznie. 
+Weźmy na warsztat przykładowy komponent:
+```typescript
+@Component({
+  selector: 'app-company',
+  template: '<div *ngIf="messageVisible" id="welcomeMessage">Hello!</div>',
+})
+export class CompanyComponent {
+  messageVisible: boolean = false;
+
+  getCompany(): Promise<string> {
+    return Promise.resolve("company");
+  }
+
+  showMessage() {
+    setTimeout(() => {
+      this.messageVisible = true;
+    }, 2000)
+  }
+}
+```
+Posiada on dwie metody: jedną, która zwraca Promise z nazwą firmy oraz drugą, która wykonuje zmianę widoczności flagi po dwóch sekundach. Na podstawie tej flagi wyświetlana jest wiadomość w templacie HTML.
+
+Klasa testowa do której będą dodawane test-case'y. W tym przypadku jest ona w 100% standardowa:
+```typescript
+describe('AppComponent', () => {
+  let fixture: ComponentFixture<AppComponent>;
+  let debugElement: DebugElement;
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        AppComponent
+      ],
+    }).compileComponents();
+          fixture = TestBed.createComponent(AppComponent);
+      debugElement = fixture.debugElement;
+  }));
+ });
+```
+
+Nie poruszę w tym wpisie jednak testowania opartego na mockowaniu/stubowaniu kodu. Jeśli jesteś zainteresowany tym tematem, zachęcam do zajrzenia do wpisu Krzyśka o [testowaniu serwisów](https://blog.consdata.tech/2019/11/20/testowanie-komponentow-i-serwisow.html).
+## Testowanie metod zwracających Promisy
+   Jeśli chcemy przetestować metodę, która zwraca wartość opakowaną w Promise oraz której wynik nie jest zależny od dostępności zewnętrznych usług, możemy w łatwy sposób sprawdzić zwracane przez nie wartości przy pomocy mechanizmu async/await:
+```typescript
+it('resolves company using async/await', async function () {
+    const company = fixture.componentInstance.getCompany();
+    expect(company).toEqual("company");
+});
```

Comment:
Tutaj przydało by się pare słów wyjaśnienia. Wspominasz o użyciu mechanizmu async/ await, ale awaita nigdzie nie ma. Kiedy oznaczymy funkcję jako async, to expect zaczyna przyjmować promisy?